Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Distinct.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT License.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Reaqtive.Operators
+{
+    internal sealed class Distinct<TResult, TKey> : SubscribableBase<TResult>
+    {
+        private readonly ISubscribable<TResult> _source;
+        private readonly Func<TResult, TKey> _keySelector;
+        private readonly IEqualityComparer<TKey> _equalityComparer;
+
+        public Distinct(ISubscribable<TResult> source, Func<TResult, TKey> keySelector, IEqualityComparer<TKey> comparer)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(keySelector != null);
+            Debug.Assert(comparer != null);
+
+            _source = source;
+            _keySelector = keySelector;
+            _equalityComparer = comparer;
+        }
+
+        protected override ISubscription SubscribeCore(IObserver<TResult> observer)
+        {
+            return new _(this, observer);
+        }
+
+        private sealed class _ : StatefulUnaryOperator<Distinct<TResult, TKey>, TResult>, IObserver<TResult>
+        {
+            private HashSet<TKey> _keySet;
+
+            public _(Distinct<TResult, TKey> parent, IObserver<TResult> observer)
+                : base(parent, observer)
+            {
+            }
+
+            protected override void OnStart()
+            {
+                if (_keySet == null)
+                {
+                    _keySet = new HashSet<TKey>(Params._equalityComparer);
+                }
+            }
+
+            public override string Name => "rc:Distinct";
+
+            public override Version Version => Versioning.v1;
+
+            public void OnCompleted()
+            {
+                Output.OnCompleted();
+                Dispose();
+            }
+
+            public void OnError(Exception error)
+            {
+                Output.OnError(error);
+                Dispose();
+            }
+
+            public void OnNext(TResult value)
+            {
+                TKey key;
+                var added = false;
+                try
+                {
+                    key = Params._keySelector(value);
+                    added = _keySet.Add(key);
+                }
+                catch (Exception exception)
+                {
+                    OnError(exception);
+                    return;
+                }
+
+                if (added)
+                {
+                    StateChanged = true;
+
+                    Output.OnNext(value);
+                }
+            }
+
+            protected override void LoadStateCore(IOperatorStateReader reader)
+            {
+                base.LoadStateCore(reader);
+
+                _keySet = reader.Read<HashSet<TKey>>();
```

Comment:
This is okay for the time being, but some future work is needed to integrate https://github.com/reaqtive/reaqtor/tree/main/Reaqtor/Pearls/OperatorLocalStorage in the core product. One of the reasons operators like `Distinct` were omitted from the initial operator set was the potentially unbounded state storage, and the fact that any little change to the state of the operator causes it to re-persist (rather than persisting deltas).

We can consider whether we want to tackle this for a 1.0 release, or later. Doing it early would be beneficial to drive out rough edges of the operator local storage feature, but it's a significant piece of work. On the other hand, doing it later results in having to deal with state migration (which is already the case for other operators like `GroupBy` which can have big storage requirements).

One thing worth considering in the meantime is to support an upper bound setting, as we've done in other places (and used in production to prevent runaway state), e.g. https://github.com/reaqtive/reaqtor/blob/main/Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/GroupBy.cs#L86. This provides a temporary middle ground where users can leverage the Reaqtive operator library as an Rx variant (and as such, we can continue to complete the operator set over here, because Reaqtive is usable by itself as well, outside the bigger Reaqtor domain), while providing a dial for those hosting it in an environment where state sizes can become problematic.

Reply:
Should we add an analyzer that gives warning to users who try to use Reaqtor v1 operators which contain exponentially growing state?